transport/conn.go

```diff
@@ -0,0 +1,222 @@
+package transport
+
+import (
+	"bufio"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"runtime"
+	"sync"
+
+	"scylla-go-driver/frame"
+	. "scylla-go-driver/frame/request"
+	. "scylla-go-driver/frame/response"
+)
+
+// TODO on send and recv i/o error we shall reset the connection
+// TODO request coelasting if there is more items in requestCh than we can send them together, we can check channel length, we need a write buffer
+
+type request struct {
+	frame.Request
+	StreamID frame.StreamID
+	Compress bool
+	Tracing  bool
+}
+
+type connWriter struct {
+	conn      io.Writer
+	buf       frame.Buffer
+	requestCh chan request
+}
+
+func (c *connWriter) submit(r request) {
+	c.requestCh <- r
+}
+
+func (c *connWriter) loop() {
+	runtime.LockOSThread()
+
+	for {
+		r, ok := <-c.requestCh
+		if !ok {
+			return
+		}
+
+		if err := c.send(r); err != nil {
+			// FIXME gracefully handle send error
+			log.Fatalf("send error: %s", err)
+		}
+	}
+}
+
+func (c *connWriter) send(r request) error {
+	c.buf.Reset()
+
+	// Dump request with header to buffer
+	h := frame.Header{
+		Version:  frame.CQLv4,
+		StreamID: r.StreamID,
+		OpCode:   r.OpCode(),
+	}
+	h.WriteTo(&c.buf)
+	r.WriteTo(&c.buf)
+	if err := c.buf.Error(); err != nil {
+		return fmt.Errorf("serialize: %w", err)
+	}
+
+	// Update length in header
+	b := c.buf.Bytes()
+	l := uint32(len(b) - frame.HeaderSize)
+	binary.BigEndian.PutUint32(b[5:9], l)
```

Comment:
How about Buffer method to update header length?

Reply:
Buffer operations are twofold, read new bytes from a stream and write (add) bytes to slice.
This is different and requires some additional assumptions.
This could be moved to a separate function in frame pkg if needed - then it can be unit tested.